File: 67.PushRelabelAlgo.py
# The Push-Relabel algorithm is a method to compute the maximum flow in a flow network.
# It maintains a preflow (may exceed the capacities temporarily) and uses push and relabel operations to send flow from the 
# source to the sink efficiently.

# Explanation

# Preflow Initialization: Start by sending the maximum 
# possible flow from the source to its neighbors (saturating outgoing edges).

# Excess Flow: Maintain an excess array, where excess[u] represents 
# how much flow is at node u waiting to be pushed further.

# Push Operation: If a vertex u has excess flow and an adjacent vertex v has a lower height, 
# push as much flow as possible from u to v.

# Relabel Operation: If u cannot push flow to any neighbor (all neighbors are 
# higher or edges are full), increase u's height to allow pushing later.

# Termination: Repeat push and relabel until no vertex (except source and sink) has excess flow.

# Push-Relabel is particularly efficient for dense graphs and has better practical 
# performance than Edmonds-Karp in many cases.

class PushRelabel:
    def __init__(self, n):
        self.n = n
        self.capacity = [[0] * n for _ in range(n)]
        self.flow = [[0] * n for _ in range(n)]
        self.height = [0] * n
        self.excess = [0] * n

    def add_edge(self, u, v, cap):
        self.capacity[u][v] = cap

    def push(self, u, v):
        push_flow = min(self.excess[u], self.capacity[u][v] - self.flow[u][v])
        self.flow[u][v] += push_flow
        self.flow[v][u] -= push_flow
        self.excess[u] -= push_flow
        self.excess[v] += push_flow
        print(f"Pushed {push_flow} from {u} to {v}")

    def relabel(self, u):
        min_height = float('inf')
        for v in range(self.n):
            if self.capacity[u][v] - self.flow[u][v] > 0:
                min_height = min(min_height, self.height[v])
        self.height[u] = min_height + 1
        print(f"Relabeled vertex {u} to height {self.height[u]}")

    def discharge(self, u):
        while self.excess[u] > 0:
            for v in range(self.n):
                if self.capacity[u][v] - self.flow[u][v] > 0 and self.height[u] == self.height[v] + 1:
                    self.push(u, v)
                    break
            else:
                self.relabel(u)

    def max_flow(self, s, t):
        self.height[s] = self.n
        for v in range(self.n):
            if self.capacity[s][v] > 0:
                self.flow[s][v] = self.capacity[s][v]
                self.flow[v][s] = -self.capacity[s][v]
                self.excess[v] = self.capacity[s][v]
                self.excess[s] -= self.capacity[s][v]
        active = [i for i in range(self.n) if i != s and i != t and self.excess[i] > 0]
        while active:
            u = active.pop(0)
            self.discharge(u)
            if self.excess[u] > 0:
                active.append(u)
        return sum(self.flow[s][i] for i in range(self.n))


pr = PushRelabel(4)
pr.add_edge(0, 1, 100)
pr.add_edge(0, 2, 100)
pr.add_edge(1, 2, 1)
pr.add_edge(1, 3, 100)
pr.add_edge(2, 3, 100)

print("Maximum flow:", pr.max_flow(0, 3))

        

File: 67.PushRelabelAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class PushRelabel {

    int n;                      // Number of vertices
    int[][] capacity;            // Capacity matrix
    int[][] flow;                // Flow matrix
    int[] height;                // Height of vertices
    int[] excess;                // Excess flow at vertices

    // Constructor to initialize graph
    public PushRelabel(int n) {
        this.n = n;
        capacity = new int[n][n];
        flow = new int[n][n];
        height = new int[n];
        excess = new int[n];
    }

    // Function to add edge with capacity
    public void addEdge(int u, int v, int cap) {
        capacity[u][v] = cap;
    }

    // Push operation from u to v
    public void push(int u, int v) {
        int pushFlow = Math.min(excess[u], capacity[u][v] - flow[u][v]);
        flow[u][v] += pushFlow;
        flow[v][u] -= pushFlow;
        excess[u] -= pushFlow;
        excess[v] += pushFlow;
        System.out.println("Pushed " + pushFlow + " from " + u + " to " + v);
    }

    // Relabel operation on vertex u
    public void relabel(int u) {
        int minHeight = Integer.MAX_VALUE;
        for (int v = 0; v < n; v++) {
            if (capacity[u][v] - flow[u][v] > 0) {
                minHeight = Math.min(minHeight, height[v]);
            }
        }
        height[u] = minHeight + 1;
        System.out.println("Relabeled vertex " + u + " to height " + height[u]);
    }

    // Discharge excess flow at vertex u
    public void discharge(int u) {
        while (excess[u] > 0) {
            boolean pushed = false;
            for (int v = 0; v < n; v++) {
                if (capacity[u][v] - flow[u][v] > 0 && height[u] == height[v] + 1) {
                    push(u, v);
                    pushed = true;
                    break;
                }
            }
            if (!pushed) {
                relabel(u);
            }
        }
    }

    // Main function to calculate maximum flow
    public int maxFlow(int s, int t) {
        height[s] = n;  // Set source height to number of vertices
        for (int v = 0; v < n; v++) {
            if (capacity[s][v] > 0) {
                flow[s][v] = capacity[s][v];
                flow[v][s] = -capacity[s][v];
                excess[v] = capacity[s][v];
                excess[s] -= capacity[s][v];
                System.out.println("Initial push from source " + s + " to " + v + " = " + capacity[s][v]);
            }
        }

        // List of active vertices (excluding source and sink)
        Queue<Integer> active = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (i != s && i != t && excess[i] > 0) active.add(i);
        }

        // Process active vertices
        while (!active.isEmpty()) {
            int u = active.poll();
            int oldExcess = excess[u];
            discharge(u);
            if (excess[u] > 0) active.add(u);
        }

        // Total flow is sum of flows leaving source
        int maxFlow = 0;
        for (int i = 0; i < n; i++) maxFlow += flow[s][i];
        return maxFlow;
    }

    // Example usage
    public static void main(String[] args) {
        PushRelabel pr = new PushRelabel(4);
        pr.addEdge(0, 1, 100);
        pr.addEdge(0, 2, 100);
        pr.addEdge(1, 2, 1);
        pr.addEdge(1, 3, 100);
        pr.addEdge(2, 3, 100);

        System.out.println("Maximum flow: " + pr.maxFlow(0, 3));
    }
}
